Make `st.table` rounded

jrieke opened the pull request:
## Describe your changes

Adds rounded borders to `st.table`, similar to other elements. This also removes the negative margin below tables. 

Before:
![CleanShot 2024-11-08 at 23 54 24@2x](https://github.com/user-attachments/assets/b3049cc3-b849-4567-822d-c11109025820)


After: 
![CleanShot 2024-11-08 at 23 53 47@2x](https://github.com/user-attachments/assets/b8e95452-2718-4943-b1a6-4b20b54de7fd)

## GitHub Issue Link (if applicable)

## Testing Plan

Updated snapshots. No new tests added since there are already pretty good e2e tests for `st.table` that cover a wide range of different tables. 

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-jrieke committed:
Make st.table rounded

sfc-gh-jrieke committed:
Formatting

sfc-gh-jrieke committed:
Add separate to eslint rule

sfc-gh-jrieke committed:
Update snapshots

lukasmasuch commented on e2e_playwright/__snapshots__/linux/st_table_test/st_table-themed[light_theme-chromium].png:
The styled cell background here goes beyond the table borders. You might also need to add this to `StyledTable`:

```typescript
  "& th:first-of-type": {
    borderTopLeftRadius: theme.radii.default,
  },
  "& th:last-of-type": {
    borderTopRightRadius: theme.radii.default,
  },
  "& tr:last-of-type td:first-of-type": {
    borderBottomLeftRadius: theme.radii.default,
  },
  "& tr:last-of-type td:last-of-type": {
    borderBottomRightRadius: theme.radii.default,
  },
```

jrieke replied:
Great catch! Had to modify the CSS a bit because it cut off some cells within the table but it should be fixed now.

lukasmasuch reviewed (approved):
LGTM 👍  the issue with the background color might need to be fixed. Also, I think the current behaviour when its scrollable is a bit suboptimal, it would be nicer if the border is applied on a wrapping container outside the table and is always fully shown (similar to st.dataframe). But thats already the case with the old version.

![image](https://github.com/user-attachments/assets/d74c40fe-b911-4d74-b61a-79ac21e872f5)

sfc-gh-jrieke committed:
Merge branch 'develop' into fix/table-roundedness

sfc-gh-jrieke committed:
Make table content rounded

sfc-gh-jrieke committed:
Fix css for corner cells

sfc-gh-jrieke committed:
Remove bottom margin

sfc-gh-jrieke committed:
Move enclosing border to a wrapper around the table

sfc-gh-jrieke committed:
Fix caption width

sfc-gh-jrieke committed:
Move to first-of-type

sfc-gh-jrieke committed:
Revert to first-child

sfc-gh-jrieke committed:
Update snapshots

sfc-gh-jrieke committed:
Add comments

jrieke commented:
> Also, I think the current behaviour when its scrollable is a bit suboptimal, it would be nicer if the border is applied on a wrapping container outside the table and is always fully shown (similar to st.dataframe).

Added this. One negative side effect of this is that the `<caption>` element needs to be placed outside of the `<table>` then to make it show outside of the border. This isn't great for accessibility. But I think in this case it's ok because adding a caption isn't a native feature and can only be done via Pandas Styler's `set_caption` method and I couldn't find a single example on GitHub where someone uses `st.table` in combination with `set_caption`. I also added a comment to the code explaining that choice.

jrieke commented on frontend/lib/src/components/elements/ArrowTable/styled-components.ts:
```diff
@@ -22,17 +22,8 @@ export const StyledTableContainer = styled.div(({ theme }) => ({
 
-export const StyledTable = styled.table(({ theme }) => ({
-  width: theme.sizes.full,
-  marginBottom: theme.spacing.lg,
```
Removed the bottom margin because all other elements also don't have top/bottom margin. Not sure if there's a specific reason for having it here.

jrieke commented:
@lukasmasuch can you give this a brief second review whenever you have time?

lukasmasuch commented on frontend/lib/src/components/elements/ArrowTable/styled-components.ts:
```diff
@@ -41,12 +32,53 @@ export const StyledTable = styled.table(({ theme }) => ({
+
+  // Add border radius to corner cells to prevent colored background from showing
+  // through the border.
+  "& thead tr:first-child th:first-child": {
+    borderTopLeftRadius: theme.radii.default,
+  },
+  "& thead tr:first-child th:last-child": {
+    borderTopRightRadius: theme.radii.default,
+  },
+  "& tbody tr:last-child td:first-child": {
+    borderBottomLeftRadius: theme.radii.default,
+  },
+  "& tbody tr:last-child td:last-child": {
+    borderBottomRightRadius: theme.radii.default,
```
This might not be needed anymore with the change to use the wrapping border

jrieke replied:
Your're right, removing.

lukasmasuch reviewed (approved):
LGTM 👍

sfc-gh-jrieke committed:
Remove redundant code

jrieke committed:
Merge branch 'develop' into fix/table-roundedness
